includes/embeds/class-amp-gallery-embed-handler.php

```diff
@@ -16,264 +15,174 @@
-					'order'          => $atts['order'],
-					'orderby'        => $atts['orderby'],
-					'fields'         => 'ids',
-				]
-			);
-		} else {
-			$attachments = get_children(
-				[
-					'post_parent'    => $id,
-					'post_status'    => 'inherit',
-					'post_type'      => 'attachment',
-					'post_mime_type' => 'image',
-					'order'          => $atts['order'],
-					'orderby'        => $atts['orderby'],
-					'fields'         => 'ids',
-				]
-			);
+		if ( $is_carousel ) {
+			$gallery_size = isset( $attrs['size'] ) ? $attrs['size'] : null;
+
+			if ( $gallery_size && 'thumbnail' === $gallery_size ) {
+				/*
+				 * If the 'gallery' shortcode has a `size` attribute of `thumbnail`, prevent outputting an <amp-carousel>.
+				 * That will often get thumbnail images around 150 x 150,
+				 * while the <amp-carousel> usually has a width of 600 and a height of 480.
+				 * That often means very low-resolution images.
+				 * So fall back to the normal 'gallery' shortcode callback, gallery_shortcode().
+				 */
+				return $html;
+			}
+
+			if ( ! $gallery_size ) {
+				// Default to `large` if no `size` attribute is specified.
+				$attrs['size'] = 'large';
+			}
 		}
 
-		if ( empty( $attachments ) ) {
-			return '';
+		if ( $is_lightbox ) {
+			// Prevent wrapping the images in anchor tags if a lightbox is specified. If that is done the link will get
+			// preference over the lightbox when the image is clicked.
+			$attrs['link'] = 'none';
 		}
 
-		$urls = [];
-		foreach ( $attachments as $attachment_id ) {
-			list( $url, $width, $height ) = wp_get_attachment_image_src( $attachment_id, $atts['size'], true );
+		// Use `data` attributes to indicate which options are configured for the embed. These indications are later
+		// processed during sanitization of the embed in `::sanitize_raw_embeds`.
+		add_filter(
+			'gallery_style',
+			static function ( $style ) use ( $is_carousel, $is_lightbox ) {
+				$data_attrs = [];
 
-			if ( ! $url ) {
-				continue;
-			}
+				if ( $is_lightbox ) {
+					$data_attrs[] = 'data-amp-lightbox';
+				}
 
-			$href = null;
-			if ( empty( $atts['lightbox'] ) && ! empty( $atts['link'] ) ) {
-				if ( 'file' === $atts['link'] ) {
-					$href = $url;
-				} elseif ( 'post' === $atts['link'] ) {
-					$href = get_attachment_link( $attachment_id );
+				if ( $is_carousel ) {
+					$data_attrs[] = 'data-amp-carousel';
 				}
+
+				return str_replace( "class='", implode( ' ', $data_attrs ) . " class='", $style );
 			}
+		);
 
-			$urls[] = [
-				'href'   => $href,
-				'url'    => $url,
-				'srcset' => wp_get_attachment_image_srcset( $attachment_id, $atts['size'] ),
-				'width'  => $width,
-				'height' => $height,
-				'alt'    => trim( wp_strip_all_tags( get_post_meta( $attachment_id, '_wp_attachment_image_alt', true ) ) ), // Logic from wp_get_attachment_image().
-				'id'     => $attachment_id,
-			];
-		}
+		remove_filter( 'post_gallery', [ $this, 'generate_gallery_markup' ] );
```

Comment:
I wonder if instead of removing and re-adding the filter, if instead it would be better to use some `static` variable to detect recursive calls.

Reply:
@pierlon What do you think of this: a6269b5c9f3163c366d9ebeb80bcd49784ba040a

Reply:
That makes it a little bit better :+1:. I wasn't totally pleased with my solution but I can get along with yours.